numba/typed/typeddict.py

```diff
@@ -260,27 +260,59 @@ def box_dicttype(typ, val, c):
 def unbox_dicttype(typ, val, c):
     context = c.context
 
-    miptr = c.pyapi.object_getattr_string(val, '_opaque')
+    # Check that `type(val) is Dict`
+    dict_type = c.pyapi.unserialize(c.pyapi.serialize_object(Dict))
+    valtype = c.pyapi.object_type(val)
+    same_type = c.builder.icmp_unsigned("==", valtype, dict_type)
 
-    mip_type = types.MemInfoPointer(types.voidptr)
-    native = c.unbox(mip_type, miptr)
+    with c.builder.if_else(same_type) as (then, orelse):
+        with then:
+            miptr = c.pyapi.object_getattr_string(val, '_opaque')
 
-    mi = native.value
+            mip_type = types.MemInfoPointer(types.voidptr)
+            native = c.unbox(mip_type, miptr)
 
-    argtypes = mip_type, typeof(typ)
+            mi = native.value
 
-    def convert(mi, typ):
-        return dictobject._from_meminfo(mi, typ)
+            argtypes = mip_type, typeof(typ)
 
-    sig = signature(typ, *argtypes)
-    nil_typeref = context.get_constant_null(argtypes[1])
-    args = (mi, nil_typeref)
-    is_error, dctobj = c.pyapi.call_jit_code(convert , sig, args)
-    # decref here because we are stealing a reference.
-    c.context.nrt.decref(c.builder, typ, dctobj)
+            def convert(mi, typ):
+                return dictobject._from_meminfo(mi, typ)
 
-    c.pyapi.decref(miptr)
-    return NativeValue(dctobj, is_error=is_error)
+            sig = signature(typ, *argtypes)
+            nil_typeref = context.get_constant_null(argtypes[1])
+            args = (mi, nil_typeref)
+            is_error, dctobj = c.pyapi.call_jit_code(convert , sig, args)
+            # decref here because we are stealing a reference.
+            c.context.nrt.decref(c.builder, typ, dctobj)
+
+            c.pyapi.decref(miptr)
```

Comment:
All I changed here is increasing the indent so these lines go inside the `if-then` branch.

Reply:
Thanks. Confirmed.